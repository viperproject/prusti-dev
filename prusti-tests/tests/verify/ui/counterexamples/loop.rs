// compile-flags: -Pcounterexample=true
// normalize-stderr-test: "final value:\s*-?[0-9]+" -> "$(FINAL_VALUE)"

use prusti_contracts::*;

#[ensures(result != 16)]
fn spurious() -> i32 {
    let mut x = 10;
    let mut y = 1;
    while (x > 0) {
        // The invariant is too weak to guarantee the postcondition even though
        // the program does.
        body_invariant!(x >= 0 && y > 0);
        x = x - 1;
        y = y + 1;
    }
    y
}

fn main() {}
