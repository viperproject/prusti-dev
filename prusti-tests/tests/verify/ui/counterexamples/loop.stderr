error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/loop.rs:6:11
   |
6  | #[ensures(result != 16)]
   |           ^^^^^^^^^^^^
   |
note: the error originates here
  --> $DIR/loop.rs:7:1
   |
7  | / fn spurious() -> i32 {
8  | |     let mut x = 10;
9  | |     let mut y = 1;
10 | |     while (x > 0) {
...  |
17 | |     y
18 | | }
   | |_^
note: counterexample for "x"
  $(FINAL_VALUE)
  --> $DIR/loop.rs:8:9
   |
8  |     let mut x = 10;
   |         ^^^^^
note: counterexample for "y"
  $(FINAL_VALUE)
  --> $DIR/loop.rs:9:9
   |
9  |     let mut y = 1;
   |         ^^^^^
note: counterexample for result
  $(FINAL_VALUE)
  --> $DIR/loop.rs:7:18
   |
7  | fn spurious() -> i32 {
   |                  ^^^

error: aborting due to previous error

